include/mbedtls/rsa.h

```diff
@@ -87,19 +97,25 @@ typedef struct
     mbedtls_mpi D;                      /*!<  private exponent  */
     mbedtls_mpi P;                      /*!<  1st prime factor  */
     mbedtls_mpi Q;                      /*!<  2nd prime factor  */
+
+    /* DP,DQ,QP are not used in NO_CRT but temporarily kept for ABI
+     * compatibility. Will be removed on next ABI changing release. */
```

Comment:
@hanno-arm - I've been meaning to highlight this for weeks, but this isn't how we deprecate changes. You need to put a `#ifdef` around the deprecated fields using an appropriate preprocessor symbol.

Reply:
@sbutcher-arm It was the intention that the deprecation this PR brings refers to the _direct usage_ of the RSA fields, as indicated in the [ChangeLog](https://github.com/ARMmbed/mbedtls/pull/1060/files#diff-02f0b547c2779d25cff89672135f20e3R23), but that it doesn't affect the _presence_ of the fields. The deprecation of direct usage it is not accompanied by `\deprecate` warnings and/or preprocessor guards because our own code of course does (and will continue to) use these fields, and we don't want to throw warnings for that. If we wanted to throw warnings for _user_ access only, we'd need to be able to tell within the header file whether it was included from a Mbed TLS compilation unit or a user provided one, for which we don't have a macro at the moment.

We could add a dedicated deprecation for the specific `DP,DQ,QP` fields in case of `NO_CRT` and remove them if `DEPRECATED_REMOVED` is set, but I don't see the benefit of this - any usage of these fields already falls under the "Don't use the RSA context fields directly" deprecation. Instead I would prefer a single clean up and change of the RSA context structure once when the deprecation period for direct structure usage expires (next API breaking changing probably).

Reply:
@hanno-arm We're only deprecating the fields in `MBEDTLS_RSA_NO_CRT` mode. They will be staying in the default configuration.

Since we don't promise to maintain ABI compatibility for non-default configurations, is there any reason not to remove the fields now, i.e. put `#if !defined(MBEDTLS_RSA_NO_CRT)` around them?

Reply:
@gilles-peskine-arm There are two different potential deprecations here:
1. Deprecating direct usage of structure fields - because users should use the API only.
That's the intended deprecation this PR brings. It is documented in the ChangeLog, but there's no simple way this could be backed up with `#ifdef's` and compile-time warnings.
2. Deprecating CRT-only structure fields if `NO_CRT` is set - because they are not needed.
The deprecation itself is already included in the first one. We _could_ add the usual `#ifdef's` in case of `NO_CRT` and remove them from the structure if `REMOVE_DEPRECATED` is defined, but as mentioned above, I'd prefer to wait until the release where direct structure usage is no longer supported and we're entirely free to restructure the context.

Reply:
Hi @hanno-arm,

I think the use of the word 'deprecate' is unintentionally confusing here. In it's strictest sense, it relates to the product and deprecation is a warning that a feature is supported now, but won't be supported in the future.

I think here we're talking about warning of a pending change on something that was never supported in the first place.

So let's look at the language, in the comment which is probably my real problem.

You state:
```
     * DP,DQ,QP are not used in NO_CRT but temporarily kept for ABI
     * compatibility. Will be removed on next ABI changing release.
```

We do not have to maintain ABI compatibility on `NO_CRT`. We only need to maintain ABI compatibility on the *default* configurations. Specifically, those configurations held in the  [`configs` directory](https://github.com/ARMmbed/mbedtls/tree/development/configs).

I don't see the need to warn about removal of these fields, nor mention ABI as that's confusing because we shouldn't be trying to maintain the ABI for this configuration. We should either remove them, or leave them in, because there's a good reason to (or deprecate them).

I'm sorry if the ABI policy was previously unknown to you. We really need a knowledgebase article discussing how we maintain the API's and ABI.

Reply:
@gilles-peskine-arm @sbutcher-arm Ok, so we either remove the comment or properly deprecate the non-CRT fields. As mentioned, I prefer the former, leaving all actual structure changes to a later stage (e.g. there was also @mpg's consideration of removing `P,Q` for non-CRT, too, and blinding with `DE-1`).

Reply:
On second thoughts this is an **API** change, not just an ABI change. We used to require DP, DQ and QP even in no-CRT mode in `mbedtls_rsa_check_privkey`, so users may have legitimately filled these fields if they were building a context manually.

Ideally we would warn users if they use these fields. We can do that with a `__attribute__((deprecated))` typedef which is deprecated only in NO_CRT mode.

Doing this *only outside of `rsa.c`* for all the fields is feasible with preprocessor tricks.

Reply:
I do think there should be a comment indicating that users should not access library. I'm indifferent as to whether the comment uses the word “deprecated” or not. It should not mention the ABI since this is not just about the ABI.

Reply:
Ok, I propose the following plan:

 1. @hanno-arm - remove the comments as they're misleading. This isn't how we deprecate fields. That's a simple fix and then I'm happy.
2. @hanno-arm - create a GitHub issue, to describe the need to properly deprecate these fields, in a way which is yet to be decided. We have partially done it in your comments, but I want to do something more robust. 
3. $$$'s! (or maybe that's another plan).

Reply:
@sbutcher-arm 1. Ok. 2 I think we want to indicate that direct access is deprecated for _all_ structure fields, not just the non-CRT ones, and find a good way to communicate that beyond the entry in the ChangeLog. I'll create a GitHub issue for that.

Reply:
@sbutcher-arm @gilles-peskine-arm Added 25b96ea to remove the comments on potential removal of the non-CRT fields.